backend/asm_targets/asm_directives.ml

```diff
@@ -0,0 +1,322 @@
+    end;
+    (* Stop dsymutil complaining about empty __debug_line sections (produces
+       bogus error "line table parameters mismatch") by making sure such
+       sections are never empty. *)
+    let file_num = A.get_file_num "none" in
+    loc ~file_num ~line:1 ~col:1;
+    D.text ()
+
+  let with_comment f ?comment x =
+    Option.iter D.comment comment;
+    f x
+
+  let ( >> ) f g x = g (f x)
+
+  let int8 =
+    with_comment (Int8.to_int >> Int64.of_int >> D.const_int64 >> D.byte)
+
+  let int16 =
+    with_comment (Int16.to_int >> Int64.of_int >> D.const_int64 >> D.word)
+
+  let int32 = with_comment (Int64.of_int32 >> D.const_int64 >> D.long)
+
+  let int64 = with_comment (D.const_int64 >> D.qword)
+
+  let uint8 =
+    with_comment (Uint8.to_int >> Int64.of_int >> D.const_int64 >> D.byte)
+
+  let uint16 =
+    with_comment (Uint16.to_int >> Int64.of_int >> D.const_int64 >> D.word)
+
+  let uint32 = with_comment (Uint32.to_int64 >> D.const_int64 >> D.long)
+
+  let uint64 = with_comment (Uint64.to_int64 >> D.const_int64 >> D.qword)
+
+  let targetint ?comment num =
+    match Targetint.repr num with
+    | Int32 n -> int32 ?comment n
+    | Int64 n -> int64 ?comment n
+
+  let uleb128 = with_comment (Uint64.to_int64 >> D.const_int64 >> D.uleb128)
+
+  let sleb128 = with_comment (D.const_int64 >> D.sleb128)
+
+  let string = with_comment (fun str -> D.bytes str)
+
+  let cache_string ?comment section str =
+    let cached : Cached_string.t = { section; str; comment } in
+    match Cached_string.Map.find cached !cached_strings with
+    | label -> label
+    | exception Not_found ->
+      let label = Asm_label.create section in
+      cached_strings := Cached_string.Map.add cached label !cached_strings;
+      label
+
+  let emit_cached_strings () =
+    Cached_string.Map.iter
+      (fun { section; str; comment } label_name ->
+        switch_to_section section;
+        define_label label_name;
+        string ?comment str;
+        int8 Int8.zero)
+      !cached_strings;
+    cached_strings := Cached_string.Map.empty
+
+  let comment str = D.comment str
+
+  let define_data_symbol symbol =
+    (* CR-someday bkhajwal: Asm_symbol currently is just a string, if
+       section-related information is added, should be worth doing this check.
+
+       check_symbol_for_definition_in_current_section symbol; *)
+    let symbol = Asm_symbol.encode symbol in
+    let data_type = D.QWORD in
+    D.label ~data_type symbol;
+    match Target_system.assembler (), Target_system.is_windows () with
+    | GAS_like, false -> D.type_ symbol "STT_OBJECT"
+    | GAS_like, true | MacOS, _ | MASM, _ -> ()
+
+  let global sym = D.global (Asm_symbol.encode sym)
+
+  let const_machine_width const = D.qword const
+
+  let symbol =
+    with_comment (fun sym ->
+        let lab = D.const_label (Asm_symbol.encode sym) in
+        const_machine_width lab)
+
+  let label ?comment:_ _lab = A.emit_line "label"
```

Comment:
A number of argument are being ignored;
my understanding is that it is on purpose
because other parts of the puzzle will be
merged later. Is there a way to be sure we
will revisit everything? Maybe by giving
`comment`, `upper`, `lower` a dummy type?

Reply:
What I had in mind was indeed to change the types
in the mli, making the functions basically unusable
by clients until everything is in place. CRs will do.